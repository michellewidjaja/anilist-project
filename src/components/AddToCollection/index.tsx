import React, { useState } from 'react';
import { Card, CardContent, 
    PageTitle, ListItem, Item, Modal, 
    ModalContent, Button, Message, Subtitle } from '../../pages/globalStyles';
import { css } from '@emotion/react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus, faXmark, faHeartCirclePlus } from '@fortawesome/free-solid-svg-icons';

import Input from '../Input';

interface CollectionListItem {
    collectionName: string,
    description: string,
    animeId: string,
    animeTitle: string,
    coverImage?: string    
}

interface Props {
    detail: {
        [key: string]: any
    }
}

interface Message {
    type?: string,
    message: string
}

export default function AddToCollection(props: Props) {
    const { detail } = props;
    const getCollectionList = localStorage.getItem('collectionList') || '';
    const parsedList = getCollectionList ? JSON.parse(getCollectionList) : {};

    const [openModal, setOpenModal] = useState(false);
    const [collectionName, setCollectionName] = useState('');
    const [message, setMessage] = useState<Message>({ message: '' });

    const handleOpenModal = () => {
        setCollectionName('');
        setMessage({ message: ''});
        setOpenModal(!openModal);
    }

    const checkIfExistInCollection = () => {
        const existingList = parsedList[collectionName] ? parsedList[collectionName] : [];
        const checkExist = existingList.find((v: CollectionListItem, k: string) => v?.animeTitle === detail.title.romaji);
        
        return checkExist?.animeTitle ? true : false;
    }

    const handleSaveCollection = () => {
        setMessage({ message: ''});

        if (collectionName === '') {
            setMessage({ message: 'Please fill in the collection name', type: 'error'});
            return false;
        } else if (/[^a-zA-Z0-9]/.test(collectionName)) {
            setMessage({ message: 'Field cannot contain special characters', type: 'error'});
            return false;
        }

        const newItem = [
            ...(parsedList[collectionName] ? parsedList[collectionName] : []),
            {
                collectionName: collectionName,
                animeId: detail.id,
                animeTitle: detail.title.romaji,
                coverImage: detail.coverImage.large
            }
        ]

        const collection = {
            ...parsedList,
            [collectionName] : newItem
        }

        if (checkIfExistInCollection()) {
            setMessage({ message: `Duplicate anime in the collection - ${collectionName}`, type: 'error'})
        } else {
            localStorage.setItem('collectionList', JSON.stringify(collection));
            setMessage({ message: `Succesfully added to ${collectionName} collection` });
        }
        setCollectionName('');
    }

    return (
        <>  
            <FontAwesomeIcon icon={faHeartCirclePlus} 
                data-testid="btnAddToCollection" 
                onClick={() => setOpenModal(true)}
                css={css`
                    color: #e95b5b;
                    font-size: 20px;
                `} 
            />
            <Modal show={openModal} data-testid="modalAddCollection">
                <ModalContent>
                    <div css={css`text-align: right`}>
                        <FontAwesomeIcon icon={faXmark} onClick={handleOpenModal}/>
                    </div>
                    <PageTitle>Add Collection</PageTitle>
                    <Card variant="gray">
                        <CardContent>
                            <Input 
                                type="text"
                                placeholder="Collection Name"
                                label="Collection Name"
                                onChange={(e: { target: { value: string }; }) => setCollectionName(e.target.value)} 
                                value={collectionName}
                                data-testid="inputCollection"
                            />
                            { message && <Message type={message.type}>{message.message}</Message> }
                        </CardContent>
                    </Card>
                    
                    <Button onClick={handleSaveCollection} data-testid="btnSaveToCollection">Save</Button>
                    <hr css={css`margin: 20px 0`} />
                    <Subtitle css={css`margin-top: 20px;`}>My Collection</Subtitle>
                    <ListItem>
                        {
                        Object.keys(parsedList).length > 0 ?
                        Object.keys(parsedList)?.map((k3: string) => {
                            return (
                                <Item key={k3} onClick={() => setCollectionName(k3)} data-testid={k3}>
                                    {k3}
                                    <FontAwesomeIcon icon={faPlus} />
                                </Item>
                            )
                        }) : 
                            <Item>There is no collection data</Item>
                        }
                    </ListItem>
                </ModalContent>
            </Modal>
        </>
    )
}