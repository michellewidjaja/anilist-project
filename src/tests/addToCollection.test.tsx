import React from 'react';
import '@testing-library/jest-dom';
import AddToCollection from '../components/AddToCollection';
import { render, screen, fireEvent } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom';
import AnimeDetailMock from './mocks/animeDetailMock';

test('renders add to collection component to match snapshot', () => {
  const tree = render(
    <MemoryRouter initialEntries={['?id=1']}>
      <AddToCollection detail={AnimeDetailMock.Media} />
    </MemoryRouter>);
  
  expect(tree).toMatchSnapshot();
});

test('show add to collection successfully', () => {
  render(
    <MemoryRouter initialEntries={['?id=1']}>
      <AddToCollection detail={AnimeDetailMock.Media} />
    </MemoryRouter>);

  const addToColl = screen.getByTestId('btnAddToCollection');
  expect(addToColl).toBeInTheDocument();
  fireEvent.click(addToColl);
  expect(screen.getByTestId('modalAddCollection')).toBeInTheDocument();
});

test('save to collection successfully',() => {
  render(
    <MemoryRouter initialEntries={['?id=1']}>
      <AddToCollection detail={AnimeDetailMock.Media} />
    </MemoryRouter>);

  const inputColl = screen.getByTestId('inputCollection');
  const btnColl = screen.getByTestId('btnSaveToCollection');
  expect(inputColl).toBeInTheDocument();
  fireEvent.change(inputColl, { target: { value: 'coll-1' } });
  fireEvent.click(btnColl);
  expect(screen.getByText('Succesfully added to coll-1 collection')).toBeInTheDocument();
  expect(inputColl).toHaveValue('');

  //save from existing collection list
  const listColl = screen.getByText('coll-1');
  fireEvent.click(listColl);
  expect(inputColl).toHaveValue('coll-1');
  fireEvent.click(btnColl);
  expect(screen.getByText('Duplicate anime in the collection - coll-1')).toBeInTheDocument();
});