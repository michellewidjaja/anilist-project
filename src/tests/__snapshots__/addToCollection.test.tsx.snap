// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders add to collection component to match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-heart-circle-plus css-b7pztv-AddToCollection"
        data-icon="heart-circle-plus"
        data-prefix="fas"
        data-testid="btnAddToCollection"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9l2.6-2.4C267.2 438.6 256 404.6 256 368c0-97.2 78.8-176 176-176c28.3 0 55 6.7 78.7 18.5c.9-6.5 1.3-13 1.3-19.6v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5zM432 512a144 144 0 1 0 0-288 144 144 0 1 0 0 288zm16-208v48h48c8.8 0 16 7.2 16 16s-7.2 16-16 16H448v48c0 8.8-7.2 16-16 16s-16-7.2-16-16V384H368c-8.8 0-16-7.2-16-16s7.2-16 16-16h48V304c0-8.8 7.2-16 16-16s16 7.2 16 16z"
          fill="currentColor"
        />
      </svg>
      <div
        class="css-1marw79"
        data-testid="modalAddCollection"
      >
        <div
          class="css-8cy5fh"
        >
          <div
            class="css-znlezv-AddToCollection"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-xmark "
              data-icon="xmark"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 384 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="css-bleihu"
          >
            Add Collection
          </div>
          <div
            class="css-1wjdaps"
          >
            <div
              class="css-6ovs0r"
            >
              <div
                class="css-10cc1hl"
              >
                <div
                  class="css-1xx0nq8"
                >
                  Collection Name
                </div>
                <input
                  data-testid="inputCollection"
                  label="Collection Name"
                  placeholder="Collection Name"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="css-lhs8c9"
              />
            </div>
          </div>
          <div
            class="css-166wyjx"
            data-testid="btnSaveToCollection"
          >
            Save
          </div>
          <hr
            class="css-1hl48p4-AddToCollection"
          />
          <div
            class="css-yt7m0t-AddToCollection"
          >
            My Collection
          </div>
          <ul
            class="css-1n0jcgx"
          >
            <li
              class="css-nlh8m7"
            >
              There is no collection data
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-heart-circle-plus css-b7pztv-AddToCollection"
      data-icon="heart-circle-plus"
      data-prefix="fas"
      data-testid="btnAddToCollection"
      focusable="false"
      role="img"
      viewBox="0 0 576 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9l2.6-2.4C267.2 438.6 256 404.6 256 368c0-97.2 78.8-176 176-176c28.3 0 55 6.7 78.7 18.5c.9-6.5 1.3-13 1.3-19.6v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5zM432 512a144 144 0 1 0 0-288 144 144 0 1 0 0 288zm16-208v48h48c8.8 0 16 7.2 16 16s-7.2 16-16 16H448v48c0 8.8-7.2 16-16 16s-16-7.2-16-16V384H368c-8.8 0-16-7.2-16-16s7.2-16 16-16h48V304c0-8.8 7.2-16 16-16s16 7.2 16 16z"
        fill="currentColor"
      />
    </svg>
    <div
      class="css-1marw79"
      data-testid="modalAddCollection"
    >
      <div
        class="css-8cy5fh"
      >
        <div
          class="css-znlezv-AddToCollection"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-xmark "
            data-icon="xmark"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="css-bleihu"
        >
          Add Collection
        </div>
        <div
          class="css-1wjdaps"
        >
          <div
            class="css-6ovs0r"
          >
            <div
              class="css-10cc1hl"
            >
              <div
                class="css-1xx0nq8"
              >
                Collection Name
              </div>
              <input
                data-testid="inputCollection"
                label="Collection Name"
                placeholder="Collection Name"
                type="text"
                value=""
              />
            </div>
            <div
              class="css-lhs8c9"
            />
          </div>
        </div>
        <div
          class="css-166wyjx"
          data-testid="btnSaveToCollection"
        >
          Save
        </div>
        <hr
          class="css-1hl48p4-AddToCollection"
        />
        <div
          class="css-yt7m0t-AddToCollection"
        >
          My Collection
        </div>
        <ul
          class="css-1n0jcgx"
        >
          <li
            class="css-nlh8m7"
          >
            There is no collection data
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
